misc: Add LULESH GPU tests

mkjost0 opened the pull request:
This adds the LULESH tests, which currently run successfully, though they still use the gem5-resources directory to build the binary used as opposed to using a pre-built binary

Change-Id: I91c511fe92b7f9d11dfb027f435573f826bc6714

mkjost0 committed:
misc: Add LULESH GPU tests

This adds the LULESH tests, which currently run successfully,
though they still use the gem5-resources directory to build the
binary used as opposed to using a pre-built binary

Change-Id: I91c511fe92b7f9d11dfb027f435573f826bc6714

mattsinc commented on .github/workflows/gpu-tests.yaml:
```diff
@@ -0,0 +1,86 @@
+        include:
+          - command-line: -j $(nproc)
+          - image: GCN3_X86
+            command-line: -j4 --ignore-style
```
Same thing about j4 here as other ones

mattsinc reviewed (commented):
Same question here about docker: I don't see where that is coming from.  Maybe the container on line 56?

abmerop reviewed (commented):
Thanks for helping to move the GPU weekly tests out of the bash script!  This looks fine for GCN3.  One thing I would like though is to test the Vega ISA as well. I realize that will take more resources.  @mattsinc and I have been planning to discuss in detail if GCN3 can be deprecated. I would also like to see Vega as part of the ALL build, which would further reduce compilation for testing, I think (not sure if each yaml file is rebuilding).

I have many more detailed thoughts on moving to Vega but it is probably better for a discussions thread rather than this PR.

mkjost0 commented:
> Thanks for helping to move the GPU weekly tests out of the bash script! This looks fine for GCN3. One thing I would like though is to test the Vega ISA as well. I realize that will take more resources. @mattsinc and I have been planning to discuss in detail if GCN3 can be deprecated. I would also like to see Vega as part of the ALL build, which would further reduce compilation for testing, I think (not sure if each yaml file is rebuilding).
> 
> I have many more detailed thoughts on moving to Vega but it is probably better for a discussions thread rather than this PR.

Yes, this is something we can look into! I think if we do that, we can put it into a separate PR so we can update all the GPU tests to use Vega at the same time after we ensure it works locally first, unless you have other thoughts on that.

BobbyRBruce committed:
misc: Merge branch 'develop' into lulesh

Change-Id: I0d1f20e77615454d5e55a643aed474e25c90195c

mattsinc commented on .github/workflows/gpu-tests.yaml:
```diff
@@ -28,6 +28,22 @@ jobs:
+    runs-on: [self-hosted, linux, x64, build]
+    container: gcr.io/gem5-test/gcn-gpu:latest
+    needs: build-gem5
+    timeout-minutes: 600 # 10 hours
```
@mkjost0 similar to other apps, how did you determine this timeout?  My recollection is that LULESH sometimes runs longer than this when running the default 10 iterations -- maybe @abmerop remembers differently though...

abmerop replied:
A full run takes around 15 hours on my machine.  I usually exit after the first iteration when testing now though. I am not sure how many this test is trying to run

mattsinc replied:
Since there are no input args to the contrary, I believe it is running the full 10 iterations: https://gem5.googlesource.com/public/gem5-resources/+/refs/heads/develop/src/gpu/lulesh/src/lulesh.hip.cc#5507

Your experience squares with mine in terms of runtime.

Harshil2107 replied:
Melissa is no longer working on the project. I don't know how she got the timeout number. How long would you recommend setting the timeout?

mattsinc replied:
I was not aware Melissa wasn't working on it anymore, sorry.

Have you tried running LULESH in your setup?  If so, how long does it take?  I would set the timeout accordingly.

abmerop replied:
24 hours should be sufficient... But I would like to discuss how redundant the run for this app is.  We don't gain much testing the default 10 iterations in this app vs. one, for example.  And with one it would take ~2 hours.

mattsinc replied:
There is an input arg to control the number of iterations, so making this change is straightforward.

In terms of why multiple -- I think there is value in at least running 2 iterations to make sure inter-kernel behavior works.  But I agree 10 is a lot.

mattsinc reviewed (dismissed):
Only one question, which may be moot if you verified.

Harshil2107 committed:
tests: updated lulesh tests to use obtain-resource

- Changed use of wget to obtain-resource

BobbyRBruce committed:
misc: Change "gpu-tests.yaml" name value

This makes the name of this different to that in "weekly-tests.yaml".

Eventually gpu-tests.yaml should be merged into "weekly-tests.yaml".

Change-Id: I071ad657202f61feb7b65d250c3ff25b134fb2b5

BobbyRBruce committed:
misc: Update name and comment for step LULESH get

Change-Id: Ibea0e2cc82e996ba569ce8900dc3b83f62390585

BobbyRBruce committed:
misc: Add missing `download-artifact` action to LULESH

Without this there's not gem5 binary to run the tests with for the
`LULESH-tests` job.

Change-Id: Idc3d444724cefdf71a7f8ad7222fce0d1b26f738

BobbyRBruce committed:
misc: Fix incorrect comment describing gpu-test workflow

Change-Id: Ia4734341306b4ae84c2a38a53b68ef3b97dda799

BobbyRBruce committed:
misc: Add spacing in LULESH job

Change-Id: Ie95640814b72bd157a11c60e6a13d293d5302fa6

BobbyRBruce committed:
misc: Increase LULESH-tests job timeout to 24 hours

During code review no one knew if 10 hours was sufficient. To ensure we
did not immediately run into timeout issues once merged, this was
increased to 24 hours.

Change-Id: I527684c2acaa467bc261c5cc127670f1a7253457

BobbyRBruce committed:
misc: Fix line length issue in src/arch/arm/regs/misc.cc

Change-Id: Id8404bc5171e02a0786c1a2eba2336f12ffc428c

BobbyRBruce committed:
misc: Merge branch 'develop' into lulesh

Change-Id: I487073922aa0d6353eb0d3f65d464d48305f92c1

BobbyRBruce committed:
misc: Remove "build" tag for "runs-on" field

We no longer tag runners in this way.

Change-Id: I57c1d86681f602f331b2738db8f577f963e3d338

mattsinc commented on src/arch/arm/regs/misc.cc:
```diff
@@ -1928,10 +1928,18 @@ faultSctlr2EL1(const MiscRegLUTEntry &entry,
         const HCRX hcrx = tc->readMiscReg(MISCREG_HCRX_EL2);
-        if (auto fault = faultHcrFgtEL1<read, g_bitfield, &HFGTR::sctlrEL1>(entry, tc, inst);
-            fault != NoFault) {
+        if (
```
I'm not sure what these ARM changes have to do with the LULESH benchmark?

BobbyRBruce replied:
I'm going to separate this out, but TL;DR I needed to add this here for now to commit. "misc.cc" breaks the line-length rules and my pre-commit was complaining. I'll let you know when it's gone

abmerop replied:
I think I have seen this when merging develop into my private branch. Thanks for fixing this!

BobbyRBruce replied:
I messed up the fix in here. I'll wait for my real fix for this in #459 to be merged to develop then fix it all.

mattsinc commented on .github/workflows/gpu-tests.yaml:
```diff
@@ -29,6 +31,37 @@ jobs:
+
+            - name: Run LULUESH tests
+              working-directory: ${{ github.workspace }}
+              run: build/GCN3_X86/gem5.opt configs/example/apu_se.py -n3 --mem-size=8GB --reg-alloc-policy=dynamic --benchmark-root="lulesh" -c lulesh
```
@BobbyRBruce : per @abmerop and I's conversation on the previous version, please use "0.01 2" as the input args and then reduce the timeout to 8 hours.

BobbyRBruce replied:
When you say passing the input args, do you mean like this?: ` build/GCN3_X86/gem5.opt configs/example/apu_se.py -n3 --mem-size=8GB --reg-alloc-policy=dynamic --benchmark-root="lulesh" -c lulesh 0.01 2` ?

mattsinc replied:
Yes, that looks correct

BobbyRBruce replied:
Done.

BobbyRBruce committed:
misc: Passed '0.01 2' to lulesh test: reduces time to 8h

Change-Id: Id0a41caf60db15253c37630d0a302da508dc8cb7

mattsinc reviewed (dismissed):
thanks!  Please let @abmerop confirm before you merge in though.

mattsinc commented on .github/workflows/gpu-tests.yaml:
```diff
@@ -29,6 +31,39 @@ jobs:
+              working-directory: ${{ github.workspace }}
+              run: |
+                  build/GCN3_X86/gem5.opt configs/example/apu_se.py -n3 --mem-size=8GB --reg-alloc-policy=dynamic --benchmark-root="lulesh" -c \
+                  lulesh 0.01 2
```
Actually, now that I look at this, did you test this?  I am surprised line 65 is "lulesh 0.01 2" instead of "0.01 2" -- maybe this is something with the new YAML format I'm not getting?

abmerop replied:
It looks right to me.  I am using the left/right view on github so the original "lulesh" (no parameters) command was moved to a new line along with adding the parameters "0.01 2".  In my opinion the other view is more confusing

mattsinc replied:
Weird, just must be my browser's view then.

abmerop commented:
> thanks! Please let @abmerop confirm before you merge in though.

done

BobbyRBruce commented:
FYI, when this is good to go (soon, I promise), i'll do a merge commit with this one so it'll be added to develop as one commit.

Just giving a heads up before people start to notice all the touch-up commits i've been doing.

BobbyRBruce committed:
misc: Merge branch 'develop' into upstream-pull-256

Change-Id: I3fc92c58174c43ba8a1017075417706357166577

BobbyRBruce committed:
misc: Merge branch 'develop' into lulesh-again

Change-Id: I42a3ac45cd0340fac7664a5d566f9e6dda4cbf2b

BobbyRBruce committed:
misc: Merge branch 'develop' into lulesh

Change-Id: I136e912fcbbb89131d153781203bb1e1bb42508f

BobbyRBruce committed:
misc: Update lulesh gpu test to use GitHub docker reg

Change-Id: I0acf09bbb2f6016b542b506dc847b20ef43e9b51

mattsinc reviewed (commented):
Sorry I could not review this faster @BobbyRBruce , I was traveling the last ~5 days.  I think the final version looks good, thanks!

Harshil2107 commented:
Hi @mattsinc,

The Lulesh tests are encountering failures due to an apparent issue with the run command. The current command reads as follows: `build/GCN3_X86/gem5.opt configs/example/apu_se.py -n3 --mem-size=8GB --reg-alloc-policy=dynamic --benchmark-root=lulesh -c lulesh 0.01 2`. Upon attempting to run the lulesh test locally, I received an error indicating that 0.01 and 2 are invalid arguments. Despite scanning through the code, I couldn't identify any section within apu_se.py that accepts these arguments. Could you provide guidance on this matter?

abmerop commented:
Hi @Harshil2107 ,  I was actually just looking at this and was wondering how this command was even working. It seems it's not.  I will post a comment on what I suspect is the issue.

abmerop commented on .github/workflows/gpu-tests.yaml:
```diff
@@ -29,6 +31,39 @@ jobs:
+            - name: Run LULUESH tests
+              working-directory: ${{ github.workspace }}
+              run: |
+                  build/GCN3_X86/gem5.opt configs/example/apu_se.py -n3 --mem-size=8GB --reg-alloc-policy=dynamic --benchmark-root="lulesh" -c \
+                  lulesh 0.01 2
```
I think this needs to be something like:

`build/GCN3_X86/gem5.opt configs/example/apu_se.py -n3 --mem-size=8GB --reg-alloc-policy=dynamic --benchmark-root="lulesh" -c lulesh --options="0.01 2"`

I suspect otherwise the 0.01 and 2 are being passed as positional arguments to the apu_se.py script

mattsinc replied:
Yes @abmerop is correct.  Those inputs are arguments to to the benchmark, not to apu_se.py.
